src/app/api/core/utils/withRetry.ts

```diff
@@ -0,0 +1,49 @@
+import { StatusableError } from '@/types/CopilotApiError'
+import pRetry, { FailedAttemptError } from 'p-retry'
+import * as Sentry from '@sentry/nextjs'
+
+export const withRetry = async <T>(fn: (...args: any[]) => Promise<T>, args: any[]): Promise<T> => {
+  let isEventProcessorRegisterd = false
+
+  return await pRetry(
+    async () => {
+      try {
+        return await fn(...args)
+      } catch (error) {
+        // Hopefully now sentry doesn't report retry errors as well. We have enough triage issues as it is
+        Sentry.withScope((scope) => {
+          if (isEventProcessorRegisterd) return
+
+          isEventProcessorRegisterd = true
+          scope.addEventProcessor((event) => {
+            if (event.level === 'error' && event.message && event.message.includes('An error occurred during retry')) {
+              return null // Discard the event as it occured during retry
+            }
+            return event
+          })
+        })
+        // Rethrow the error so pRetry can rety
+        throw error
+      }
+    },
+
+    {
+      retries: 3,
+      minTimeout: 500,
+      maxTimeout: 500,
+      factor: 2, // Exponential factor for timeout delay. Tweak this if issues still persist
+      onFailedAttempt: (error: FailedAttemptError) => {
+        console.warn(
+          `CopilotAPI#withRetry | Attempt ${error.attemptNumber} failed. There are ${error.retriesLeft} retries left. Error:`,
+          error,
+        )
+      },
+      shouldRetry: (error: any) => {
+        // Typecasting because Copilot doesn't export an error class
+        const err = error as StatusableError
```

Comment:
We are casting the error to `StatusableError` to check the status. This makes sense given the context, but doesn't ensure that the cast is safe and handled, especially in production where an unexpected error type could occur.

Reply:
This shouldn't be an issue because only CopilotApiError will be caught here. I'll rework the typecasting to make this more verbose

Reply:
We are returning error.status here. If  error does not satisfy StatusableError, `error.status === 429` will simply return a falsey value. Type assumption is completely safe in use cases like this.